tutorials/llm/llama/domain-adaptive-pretraining/code/extend_tokenizer_utils.py

```diff
@@ -0,0 +1,286 @@
+            data_root (str): Path to general/domain specific data to identify tokens and extend tokenizer
+            original_tokenizer_path (str): Path to original tokenizer (llama 2 tokenizer downlaoded from hf)
+            new_tokens (List(str)): List of idenitfied high frequency tokens
+            new_vocab_path (str): Path to new vocabulary file
+            new_model_path (str): Path to new/customized tokenizer
+            old_ebd_path (str): Path to original llama2 embedding weights downlaoded from hf
+            new_ebd_path (str): Path to new embedding weights (modified due to tokenizer changes)
+            split (int): Number of splits used for original model weights (model parallelism)
+
+    Returns:
+           New model files created and saved in the paths specified below
+
+    """
+    m = model.ModelProto()
+    m.ParseFromString(open(original_tokenizer_path, 'rb').read())
+    ori_vocab_size = len(m.pieces)
+
+    print("token_cnt with original tokenizer: ")
+    sp = spm.SentencePieceProcessor()
+    sp.load(original_tokenizer_path)
+    get_token_cnt_spm(data_root, sp, batchsize=1000, keys=["text"])
+
+    add_normal_cnt = len(new_tokens)
+    add_dummy_cnt = (len(new_tokens) // 1024 + 1) * 1024 - len(new_tokens)
+    total_add_cnt = add_normal_cnt + add_dummy_cnt
+    new_vocab_size = total_add_cnt + ori_vocab_size
+    total_cnt = new_vocab_size + 768  ## consider 768 padding vocab in llama/mixtral tokenizer
+    print("original vocab_size: ", ori_vocab_size)
+    print("added normal vocab: ", add_normal_cnt)
+    print("added dummy vocab: ", add_dummy_cnt)
+    print("new vocab_size: ", new_vocab_size)
+    print("padded vocab: ", 768)
+    print("total cnt (with padding vocab): ", total_cnt)
+    assert add_dummy_cnt >= 3, "there should be at least 3 extra tokens for finetuning"
+
+    record = []
+    N = len(m.pieces)
+    for i, sym in enumerate(new_tokens):
+        new_sym = m.SentencePiece()
+        new_sym.piece = sym
+        new_sym.score = 0.0  # default score for USER_DEFINED
+        new_sym.type = 4  # type value for USER_DEFINED
+        m.pieces.insert(N + i, new_sym)  # position after default control symbols ("<unk>", "<s>", "</s>")
+        record.append([sym, N + i])
+
+    N = len(m.pieces)
+    # extra_token_id = 3
+    # for i in range(total_cnt - N):
+    for i in range(add_dummy_cnt):
+        new_sym = m.SentencePiece()
+        new_sym.piece = f"<extra_id_{i}>"
+        new_sym.score = 0.0  # default score for USER_DEFINED
+        new_sym.type = 4  # type value for USER_DEFINED
+        m.pieces.insert(N + i, new_sym)  # position after default control symbols ("<unk>", "<s>", "</s>")
+        # extra_token_id += 1
+        record.append([new_sym.piece, N + i])
+
+    with open(new_vocab_path, "w", encoding="utf8") as fp:
+        json.dump(record, fp)
+
+    with open(new_model_path, 'wb') as f:
+        f.write(m.SerializeToString())
+
+    print("token_cnt with customized tokenizer: ")
```

Comment:
The rest of function from here on will expand the embedding&output layer by append to original embedding/output, and then save to disk. However: 1. the tokens we obtain here are not filtered/sorted by frequency yet. 2. there is similar function in `extend_tokenizer_utils.py:extend_tokenizer_high_freq_tokens` that will save the extended embedding for high freq tokens again and write to the same file. Should this part be deleted?